crypto/txsigner_eip155.go

```diff
@@ -2,90 +2,166 @@ package crypto
+//
+// EIP155Signer accepts the following types of transactions:
+//   - EIP-155 replay protected transactions, and
+//   - pre-EIP-155 legacy transactions
+func NewEIP155Signer(chainID uint64) *EIP155Signer {
 	return &EIP155Signer{
-		chainID:     chainID,
-		isHomestead: isHomestead,
+		chainID: chainID,
 	}
 }
 
-// Hash is a wrapper function that calls calcTxHash with the EIP155Signer's chainID
-func (e *EIP155Signer) Hash(tx *types.Transaction) types.Hash {
-	return calcTxHash(tx, e.chainID)
-}
+// Hash returns the keccak256 hash of the transaction
+//
+// The EIP-155 transaction hash preimage is as follows:
+// RLP(nonce, gasPrice, gas, to, value, input, chainId, 0, 0)
+//
+// Specification: https://eips.ethereum.org/EIPS/eip-155#specification
+func (signer *EIP155Signer) Hash(tx *types.Transaction) types.Hash {
+	if tx.ChainID() == nil || tx.ChainID().Cmp(big.NewInt(0)) == 0 {
+		return signer.HomesteadSigner.Hash(tx)
+	}
 
-// Sender returns the transaction sender
-func (e *EIP155Signer) Sender(tx *types.Transaction) (types.Address, error) {
-	protected := true
+	var hash []byte
 
-	// Check if v value conforms to an earlier standard (before EIP155)
-	bigV := big.NewInt(0)
+	RLP := arenaPool.Get()
 
-	v, r, s := tx.RawSignatureValues()
-	if v != nil {
-		bigV.SetBytes(v.Bytes())
+	// RLP(-, -, -, -, -, -, -, -, -)
+	hashPreimage := RLP.NewArray()
+
+	// RLP(nonce, -, -, -, -, -, -, -, -)
+	hashPreimage.Set(RLP.NewUint(tx.Nonce()))
+
+	// RLP(nonce, gasPrice, -, -, -, -, -, -, -)
+	hashPreimage.Set(RLP.NewBigInt(tx.GasPrice()))
+
+	// RLP(nonce, gasPrice, gas, -, -, -, -, -, -)
+	hashPreimage.Set(RLP.NewUint(tx.Gas()))
+
+	// Checking whether the transaction is a smart contract deployment
+	if tx.To() == nil {
+		// RLP(nonce, gasPrice, gas, to, -, -, -, -, -)
+		hashPreimage.Set(RLP.NewNull())
+	} else {
+		// RLP(nonce, gasPrice, gas, to, -, -, -, -, -)
+		hashPreimage.Set(RLP.NewCopyBytes((*(tx.To())).Bytes()))
 	}
 
-	if vv := bigV.Uint64(); bits.Len(uint(vv)) <= 8 {
-		protected = vv != 27 && vv != 28
+	// RLP(nonce, gasPrice, gas, to, value, -, -, -, -)
+	hashPreimage.Set(RLP.NewBigInt(tx.Value()))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, -, -, -)
+	hashPreimage.Set(RLP.NewCopyBytes(tx.Input()))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, chainId, -, -)
+	hashPreimage.Set(RLP.NewUint(signer.chainID))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, chainId, 0, -)
+	hashPreimage.Set(RLP.NewUint(0))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, chainId, 0, 0)
+	hashPreimage.Set(RLP.NewUint(0))
+
+	// keccak256(RLP(nonce, gasPrice, gas, to, value, input))
+	hash = keccak.Keccak256Rlp(nil, hashPreimage)
+
+	arenaPool.Put(RLP)
+
+	return types.BytesToHash(hash)
+}
+
+// Sender returns the sender of the transaction
+func (signer *EIP155Signer) Sender(tx *types.Transaction) (types.Address, error) {
+	if tx.ChainID() == nil || tx.ChainID().Cmp(big.NewInt(0)) == 0 {
```

Comment:
This is a strange condition that we didn't have before, and to me is not correct.

Here we should check if the transaction is `Legacy`. If it's not we should return an error.
Then, we are missing the `isProtected` check. If it is not protected, `Homstead` signer should be called.
If it is protected, then do the check about chain id, if it is the same as we have, something lilke:

```go
if tx.ChainId().Cmp(s.chainId) != 0 {
     return common.Address{}, fmt.Errorf("invalid chain id for signer: have %d want %d", tx.ChainId(), signer.chainId)
}
```

if chain id is as expected, go ahead with the rest of the logic.

Why is this so? Because, if we have EIP155 fork enabled, we will always need to check chain id (because of replay attacks), and not go straight to `Homstead` signer if transaction did not include chain id.